src/controllers/plugins/PluginController.cpp

```diff
@@ -0,0 +1,318 @@
+{
+    qCDebug(chatterinoLua) << "Running lua file" << index;
+    lua_State *l = luaL_newstate();
+    PluginController::openLibrariesFor(l, meta);
+
+    auto pluginName = pluginDir.dirName();
+    auto plugin = std::make_unique<Plugin>(pluginName, l, meta, pluginDir);
+
+    for (const auto &[codename, other] : this->plugins_)
+    {
+        if (other->meta.name == meta.name)
+        {
+            plugin->isDupeName = true;
+            other->isDupeName = true;
+        }
+    }
+    this->plugins_.insert({pluginName, std::move(plugin)});
+    if (!PluginController::isEnabled(pluginName))
+    {
+        qCInfo(chatterinoLua) << "Skipping loading" << pluginName << "("
+                              << meta.name << ") because it is disabled";
+        return;
+    }
+
+    int err = luaL_dofile(l, index.absoluteFilePath().toStdString().c_str());
+    if (err != 0)
+    {
+        qCWarning(chatterinoLua)
+            << "Failed to load" << pluginName << "plugin from" << index << ": "
+            << lua::humanErrorText(l, err);
+        return;
+    }
+    qCInfo(chatterinoLua) << "Loaded" << pluginName << "plugin from" << index;
+}
+
+bool PluginController::reload(const QString &codename)
+{
+    auto it = this->plugins_.find(codename);
+    if (it == this->plugins_.end())
+    {
+        return false;
+    }
+    if (it->second->state_ != nullptr)
+    {
+        lua_close(it->second->state_);
+        it->second->state_ = nullptr;
+    }
+    for (const auto &[cmd, _] : it->second->ownedCommands)
+    {
+        getApp()->commands->unregisterPluginCommand(cmd);
+    }
+    it->second->ownedCommands.clear();
+    if (PluginController::isEnabled(codename))
+    {
+        QDir loadDir = it->second->loadDirectory_;
+        this->plugins_.erase(codename);
+        this->tryLoadFromDir(loadDir);
+    }
+    return true;
+}
+
+void PluginController::callEvery(const QString &functionName)
+{
+    for (const auto &[name, plugin] : this->plugins_)
+    {
+        lua_getglobal(plugin->state_, functionName.toStdString().c_str());
+        lua_pcall(plugin->state_, 0, 0, 0);
+    }
+}
+
+void PluginController::callEveryWithArgs(
+    const QString &functionName, int count,
+    const std::function<void(const std::unique_ptr<Plugin> &pl, lua_State *L)>
+        &argCb)
+{
+    for (const auto &[name, plugin] : this->plugins_)
+    {
+        lua_getglobal(plugin->state_, functionName.toStdString().c_str());
+        argCb(plugin, plugin->state_);
+        lua_pcall(plugin->state_, count, 0, 0);
+    }
+}
+
+QString PluginController::tryExecPluginCommand(const QString &commandName,
+                                               const CommandContext &ctx)
+{
+    for (auto &[name, plugin] : this->plugins_)
+    {
+        if (auto it = plugin->ownedCommands.find(commandName);
+            it != plugin->ownedCommands.end())
+        {
+            const auto &funcName = it->second;
+
+            auto *L = plugin->state_;  // NOLINT
```

Comment:
```suggestion
            auto *L = plugin->state_;
```

This can be removed, if `L` is added as ignored (which should be fine):

```yaml
# .clang-tidy
CheckOptions:
  - key: readability-identifier-naming.LocalPointerIgnoredRegexp
    value: ^L$ # Lua state
```